File: debop4s-data/src/test/scala/debop4s/data/tests/mapping/inheritance/JoinedSubclassTest.scala
package debop4s.data.tests.mapping.inheritance

import debop4s.core.utils.Hashs
import debop4s.data.model.LongEntity
import debop4s.data.tests.AbstractJpaTest
import java.util
import javax.persistence._
import org.hibernate.annotations.CacheConcurrencyStrategy
import org.hibernate.{annotations => hba}
import org.junit.Test

/**
 * 부모 클래스를 공통으로 사용하고, 자식 클래스들은 각각의 고유의 정보를 각자 테이블에 저장합니다.
 * subclass 와는 달리 자식 클래스들이 not null 속성을 가집니다.
 * 단 MySQL 같이 join 성능이 않좋은 RDBMS에서는 사용을 자제해야 합니다.
 *
 * Created by debop on 2014. 3. 9.
 */
@org.springframework.transaction.annotation.Transactional
class JoinedSubclassTest extends AbstractJpaTest {

    @PersistenceContext val em: EntityManager = null

    @Test
    def joinedSubclass() {
        val employee = new JoinedSubclassEmployee()
        employee.empNo = "21011"
        employee.name = "배성혁"
        employee.regidentNo = "111111-1111111"

        val customer = new JoinedSubclassCustomer()
        customer.name = "customer1"
        customer.contactEmployee = employee
        customer.regidentNo = "222222-2222222"
        customer.mobile = "010-1111-2222"

        em.persist(employee)
        em.persist(customer)
        em.flush()
        em.clear()

        val customer1 = em.find(classOf[JoinedSubclassCustomer], customer.id)
        assert(customer1 == customer)

        val employee1 = em.find(classOf[JoinedSubclassEmployee], employee.id)
        assert(employee1 == employee)

        em.remove(customer1)
        em.remove(employee1)
        em.flush()

        assert(em.find(classOf[JoinedSubclassCustomer], customer.id) == null)
        assert(em.find(classOf[JoinedSubclassEmployee], employee.id) == null)
    }

}

@Entity
@Inheritance(strategy = InheritanceType.JOINED)
abstract class JoinedSubclassPerson extends LongEntity {

    @Column(name = "personName", nullable = false, length = 128)
    @Index(name = "ix_joinedsubclassperson_name", columnList = "name, regidentNo")
    var name: String = _

    @Column(nullable = false, length = 128)
    var regidentNo: String = _

    override def hashCode(): Int = Hashs.compute(name, regidentNo)
}

@Entity
@org.hibernate.annotations.Cache(region = "inheritance", usage = CacheConcurrencyStrategy.READ_WRITE)
@hba.DynamicInsert
@hba.DynamicUpdate
class JoinedSubclassCustomer extends JoinedSubclassPerson {

    @Column(nullable = false)
    @Index(name = "ix_joinedsubclasscustomer_mobile", columnList = "mobile")
    var mobile: String = _

    @ManyToOne
    @JoinColumn(name = "contactEmployeeId", nullable = false)
    var contactEmployee: JoinedSubclassEmployee = _

    override def hashCode(): Int = Hashs.compute(super.hashCode(), mobile)
}

@Entity
@org.hibernate.annotations.Cache(region = "inheritance", usage = CacheConcurrencyStrategy.READ_WRITE)
@hba.DynamicInsert
@hba.DynamicUpdate
class JoinedSubclassEmployee extends JoinedSubclassPerson {

    @Column(name = "empNo", nullable = false)
    @Index(name = "ix_joinedsubclassemployee_empno", columnList = "empNo")
    var empNo: String = _

    @ManyToOne
    @JoinColumn(name = "managerId")
    var manager: JoinedSubclassEmployee = _

    @OneToMany(mappedBy = "manager", cascade = Array(CascadeType.ALL))
    @hba.LazyCollection(hba.LazyCollectionOption.EXTRA)
    var members: util.Set[JoinedSubclassEmployee] = new util.HashSet[JoinedSubclassEmployee]

    override def hashCode(): Int = Hashs.compute(super.hashCode(), empNo)
}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/inheritance/joinedsubclass/JoinedSubclassTest.java
package debop4s.data.orm.jtests.mapping.inheritance.joinedsubclass;

import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.data.orm.s.mapping.inheritance.joinedsubclass.JoinedSubclassJUnitSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 7. 22. 오후 5:34
 */
@Slf4j
@Transactional
public class JoinedSubclassTest extends JpaTestBase {

    @PersistenceContext
    EntityManager em;

    @Test
    public void personTest() {
        JoinedSubclassEmployee employee = new JoinedSubclassEmployee();
        employee.setEmpNo("21011");
        employee.setName("배성혁");
        employee.setRegidentNo("111111-11111111");

        JoinedSubclassCustomer customer = new JoinedSubclassCustomer();
        customer.setName("customer1");
        customer.setContactEmployee(employee);
        customer.setRegidentNo("222222-22222222");

        em.persist(employee);
        em.persist(customer);
        em.flush();

        em.clear();

        JoinedSubclassCustomer customer1 = em.find(JoinedSubclassCustomer.class, customer.getId());
        assertThat(customer1).isEqualTo(customer);
        assertThat(customer1.getRegidentNo()).isEqualTo(customer.getRegidentNo());

        JoinedSubclassEmployee employee1 = em.find(JoinedSubclassEmployee.class, employee.getId());
        assertThat(employee1).isEqualTo(employee);
        assertThat(employee1.getRegidentNo()).isEqualTo(employee.getRegidentNo());

        customer1 = em.find(JoinedSubclassCustomer.class, customer.getId());
        assertThat(customer1).isEqualTo(customer);
        assertThat(customer1.getRegidentNo()).isEqualTo(customer.getRegidentNo());

        employee1 = em.find(JoinedSubclassEmployee.class, employee.getId());
        assertThat(employee1).isEqualTo(employee);
        assertThat(employee1.getRegidentNo()).isEqualTo(employee.getRegidentNo());

        em.remove(customer1);
        em.remove(employee1);
        em.flush();

        assertThat(em.find(JoinedSubclassCustomer.class, customer.getId())).isNull();
        assertThat(em.find(JoinedSubclassEmployee.class, employee.getId())).isNull();
    }
}
